File: linkedList.java
//linked list with dummy node cycle mode
import java.util.Iterator;

public class LinkedLinearList<T> implements Iterable<LinkedLinearList<T>> {
    private Node<T> headerNode;
    private int mSize;

    public LinkedLinearList() {
        mSize = 0;
        headerNode = new Node<T>();
        headerNode.next = headerNode;
        headerNode.prev = headerNode;
    }

    @Override
    public Iterator<LinkedLinearList<T>> iterator() {
        return new LinkedListForwardIterator<>();
    }

    public Iterator<T> ForwardIterator() {
        return new LinkedListForwardIterator(headerNode);
    }

    public Iterator<T> BackwardIterator() {
        return new LinkedListBackwardIterator(headerNode);
    }

    // The returned iterator would point to the inserted element.
    public void insert(int position, T data) throws IndexOutOfBoundsException {
        if (position > size()) {
            throw new IndexOutOfBoundsException();
        }
        Node<T> newnode = new Node<>();
        newnode.data = data;

        Node<T> lastnode = headerNode;
        for (int i = 0; i < position; i++) {
            lastnode = lastnode.next;
        }
        newnode.prev = lastnode;
        newnode.next = lastnode.next;
        lastnode.next.prev = newnode;
        lastnode.next = newnode;
        mSize++;
    }

    // The returned iterator would point to the element after the removed one.
    public void remove(int position) throws IndexOutOfBoundsException {
        if (position > size()) {
            throw new IndexOutOfBoundsException();
        }
        Node<T> lastnode = headerNode;
        for (int i = 0; i < position; i++) {
            lastnode = lastnode.next;
        }
        lastnode.next.next.prev = lastnode;
        lastnode.next = lastnode.next.next;
        mSize--;
    }

    public int size() {
        return mSize;
    }

    @Override
    public String toString() {
        String str = "";
        Node currNode = headerNode;
        for (int i = 0; i < mSize; i++) {
            currNode = currNode.next;
            str += (" " + currNode.data);
        }
        return str;
    }

    private class Node<T> {
        public T data;
        public Node prev;
        public Node next;
    }

    public class LinkedListForwardIterator<T> implements Iterator<T> {
        private Node<T> currNode;

        public LinkedListForwardIterator() {
            this.currNode = null;
        }

        public LinkedListForwardIterator(Node<T> node) {
            this.currNode = node;
        }

        public T retrieve() {//it's get equivalent
            return this.currNode.data;
        }

        @Override
        public T next() { // preincrement
            currNode = currNode.next;
            return currNode.data;
        }

        @Override
        public boolean hasNext() {
            return currNode.next != headerNode;
        }
    }

    public class LinkedListBackwardIterator<T> implements Iterator<T> {
        Node<T> currNode;

        public LinkedListBackwardIterator() {
            this.currNode = null;
        }

        public LinkedListBackwardIterator(Node<T> node) {
            this.currNode = node;
        }

        public T retrieve() {//it's get equivalent
            return this.currNode.data;
        }

        @Override
        public T next() { // preincrement
            currNode = currNode.prev;
            return currNode.data;
        }

        @Override
        public boolean hasNext() {
            return currNode.prev != headerNode;
        }
    }
}

File: linkedList.cpp
//linked list with dummy node cycle mode
#pragma once
#include <ostream>

template <class T>
class LinkedLinearList{
private:
		class Node{
		public:
			T	mData;
			Node*	mPrev;
			Node*	mNext;
		};

public:
	class Iterator{
	public:
		//constructors:
		Iterator ()
			:mCurNode (NULL){}

		Iterator (const Iterator &other)
			:mCurNode (other.mCurNode){}
		
		Iterator (Node* node)
			:mCurNode (node){}
		
		//overloading operators:
		virtual Iterator& operator++(){ // preincrement
			mCurNode = mCurNode->mNext;
			return *this;
		}

		virtual Iterator operator++(int dummy){ // postincrement
			Iterator temp = Iterator(mCurNode);
			mCurNode = mCurNode->mNext;
			return temp;
		}

		virtual Iterator& operator--(){ // predecrement
			mCurNode = mCurNode->mPrev;
			return *this;
		}

		virtual Iterator operator--(int dummy){ // postdecrement
			Iterator temp = Iterator(mCurNode);
			mCurNode = mCurNode->mPrev;
			return temp;
		}

		bool operator!=(const Iterator& right)const {
			return (mCurNode) != ((right.mCurNode));
		}

		bool operator==(const Iterator& right)const {
			return (mCurNode) == ((right.mCurNode));
		}

		T* operator->() const {
			return &(mCurNode->mData);
		}

		T& operator*() const {
			return (mCurNode->mData);
		}

		friend class LinkedLinearList;
	private:
		//const LinkedLinearList<T> *mLinkedList;
		Node *mCurNode;
	};

	LinkedLinearList(void){
		mSize = 0;
		mHeaderNode = new Node();
		mHeaderNode->mNext = mHeaderNode;
		mHeaderNode->mPrev = mHeaderNode;
	}

	virtual ~LinkedLinearList(void) {
		Node *n, *nn; 
		for (n = mHeaderNode->mNext; n != mHeaderNode;){
			nn = n->mNext;
			delete n;
			n = nn;
		}
		delete mHeaderNode;
	}

public:
	virtual Iterator begin() const {
		return Iterator(mHeaderNode->mNext);
	}

	virtual Iterator end() const {
		return Iterator(mHeaderNode);
	}

	virtual Iterator rbegin() const {
		return Iterator(mHeaderNode->mPrev);
	}

	virtual Iterator rend() const {
		return Iterator(mHeaderNode);
	}

	// The returned iterator would point to the inserted element.
	virtual Iterator insert(Iterator it, const T& data) {
		Node *newnode = new Node();
		newnode->mData = data;
		Node *lastnode = (it.mCurNode);
		newnode->mPrev = lastnode;
		newnode->mNext = lastnode->mNext;
		lastnode->mNext->mPrev = newnode;
		lastnode->mNext = newnode;
		it++;
		mSize++;
		return it;
	}

	// The returned iterator would point to the element after the removed one.
	virtual Iterator remove(Iterator it) {
		Node *curnode = (it.mCurNode);
		curnode->mPrev->mNext = curnode->mNext;
		curnode->mNext->mPrev = curnode->mPrev;

		it++;
		mSize--;

		delete curnode;
		return it;
	}

	virtual int size() const {
		return mSize;
	};

private:
	Node* mHeaderNode;
	int mSize;
};

template <class T>
std::ostream& operator<<(std::ostream& os, const LinkedLinearList<T>& ll){
	LinkedLinearList<T>::Iterator itr;

	for (itr = ll.begin(); itr != ll.end(); itr++)
		os << *itr << "\t";
	os << "\n";
	return os;
}
